syntax = "proto3";

package leilao;

// --- Mensagens de Dados ---

// Representa os dados completos de um leilão.
message LeilaoData {
    int32 id = 1;
    string descricao = 2;
    float valor_inicial = 3;
    string inicio = 4; // Formato ISO string
    string fim = 5;    // Formato ISO string
    string status = 6; // "pendente", "ativo", "encerrado"
    float valor_atual = 7;
}

// Mensagem genérica para eventos em tempo real (SSE).
// Substitui o JSON livre que antes trafegava no RabbitMQ.
message Evento {
    string tipo = 1; // "leilao_iniciado", "lance_validado"
    string payload_json = 2; // Conteúdo do evento serializado em JSON
}

message Empty {}

// --- Definição dos Serviços (Interfaces gRPC) ---

service AuctionService {
    // Cria um leilão e retorna seus dados.
    rpc CriarLeilao (LeilaoData) returns (LeilaoData);
    
    // Retorna um fluxo (stream) de leilões ativos (Server Streaming).
    rpc GetLeiloesAtivos (Empty) returns (stream LeilaoData);
    
    // Mantém um canal aberto enviando eventos do MS Leilão para o Gateway.
    rpc SubscribeEventos (Empty) returns (stream Evento);
}